### PR TITLE
adds filter_edges and filter_vertices

### PR DESCRIPTION
**Pull request by sbromberger**

### CODE REVIEW

**Comment by kescobo** [comment]
Glorious!

**Comment by sbromberger** [comment]
@kescobo - now with induced subgraphs. Check this out:
```
filt = filter_vertices(g, :color, "red") # select all the red nodes
collect(filt) # which vertices are red?
g2 = g[filt] # give me a graph of just the red nodes and the links between them
```

**Comment by kescobo** [comment]
@sbromberger Cool! But I'm a bit confused - is `filter_vertices!()` actually modifying `g`? If yes, why do you need the second step? And if not, why the `!` in the function signature?

**Comment by kescobo** [comment]
Might be cool to have a `Range`-like object that's returned from a call to `filter_X()`, that can be used to index or make a view with the syntax in line 2 there.

**Comment by jpfairbanks** [comment]
This is awesome!

Maybe we need domain filtering too. So you can do filtering in the vertex domain rather than the property domain.

**Comment by sbromberger** [comment]
Now with generic functions:
```
f = filter_edges(mg, (g, e) -> has_prop(g, e, :weight) && get_prop(g, e, :weight) > 0.3)
```

What this filter does is left as an exercise to the reader :)

**Comment by sbromberger** [comment]
>  is filter_vertices!() actually modifying g? If yes, why do you need the second step. And if not, why the ! in the function signature?

Oops. That was a typo. Non-mutating.

**Comment by jpfairbanks** [comment]
Yeah `filter_vertices` creates an indexing object and then the `getindex` creates the subgraph.

**Comment by sbromberger** [comment]
@kescobo 

> Might be cool to have a Range-like object that's returned from a call to filter_X, that can be used to index or make a view with the syntax in line 2 there.

Calling `induced_subgraph` directly gives you the map (it returns a vector of "old vertex ID" indexed by "new vertex ID").

File: src/MetaGraphs.jl [hunk 1]
```diff
@@ -31,7 +32,12 @@ export
     rem_prop!,
     has_prop,
     clear_props!,
-    set_weightfield!
+    set_weightfield!,
+    set_defaultweight!,
```
Comment (jpfairbanks) [inline src/MetaGraphs.jl]: I am a little worried about this JAVA style interface, why do you need getters and setters? They just get and set the actual fields.

what about using the names `defaultweight(g)` for the getter and `defaultweight!(g, value)` for the setter?
You still get the encapsulation you want by having the function call in there, but you don't make it read like ☕️ ☕️ ☕️ ☕️ ☕️ ☕️ ☕️

Comment (sbromberger) [reply src/MetaGraphs.jl]: I am trying to make explicit when it's OK to access an attribute in a struct. I agree it sucks. I like your idea, though, and I'm going to do that throughout now.

Comment (sbromberger) [reply src/MetaGraphs.jl]: and I hate java. :)

**Comment by sbromberger** [comment]
> Maybe we need domain filtering too. So you can do filtering in the vertex domain rather than the property domain.

We have that now, right?

**Comment by jpfairbanks** [comment]
But does it respect the metadata? When you renumber the vertices (because you removed some), then you have to renumber the edge metadata dictionaries.

Unless you just leave holes for what you deleted.

**Comment by sbromberger** [comment]
> But does it respect the metadata? When you renumber the vertices (because you removed some), then you have to renumber the edge metadata dictionaries.

Yes. [Check out `_copy_props!()`](https://github.com/JuliaGraphs/MetaGraphs.jl/pull/2/files#diff-db67aeded3828cb429747d5266da997cR297).

Note that `filter_*` just creates a (lazy) filter iterator. It doesn't make any graph remappings until you apply it via `induced_subgraph`; if you just call `collect` on it, it will return a list of the *current* vertices/edges.

**Comment by jpfairbanks** [comment]
Oh yeah that does what I was thinking.

**Comment by sbromberger** [comment]
I'm going to go ahead and merge this - we can fix things up later.

**Pull request merged**